### PR TITLE
Guard silently-ignored GRPO config options and remove dead batch chunking

### PR DESCRIPTION
**Pull request by danielhanchen**
## Summary

Two robustness fixes to the optimized GRPO loss path in `unsloth_zoo/rl_replacements.py` (`grpo_compute_loss`, `grpo_accumulated_loss`, `UnslothEfficientGRPO`), found in a GRPO loss audit. Each was confirmed against the code and TRL's `GRPOConfig` before fixing.

## Issue 1: silently ignored GRPO config options

`UnslothGRPOConfig` forwards every TRL `GRPOConfig` field through `**kwargs`, but the optimized path does not implement two of them:

- `top_entropy_quantile < 1.0` (TRL entropy masking from "Beyond the 80/20 Rule": only the top-quantile-entropy tokens are backpropagated).
- `use_bias_correction_kl = True` (DeepSeek-V3.2: multiplies the KL term by the importance-sampling ratio).

Neither `top_entropy_quantile` nor `use_bias_correction_kl` appears anywhere in `unsloth_zoo`, so a user who set either got plain GRPO with no error and no effect.

**Fix:** a new warn-once helper `_warn_unsupported_grpo_options(trainer)`, called at the start of `grpo_accumulated_loss`, emits a clear one-time warning (guarded by a `trainer._unsloth_grpo_unsupported_warned` flag, so at most once per trainer) when an unsupported option is set to a non-default value (after Issue 3, only `top_entropy_quantile`). It uses `unsloth_zoo`'s `logger` with a `warnings.warn` fallback. Defaults are read from TRL's `GRPOConfig`: `top_entropy_quantile` defaults to `1.0` in every version checked (0.27.0 through 1.12.0), so only non-default values warn. The features are not implemented, only surfaced. The helper is imported inside `grpo_accumulated_loss` (not at module scope) so the copy of that function inlined into the generated `UnslothGRPOTrainer` cache resolves it from the installed package.

## Issue 2: dead and incorrect batch chunking

`grpo_accumulated_loss` snapped the user-facing `n_chunks` (from `GRPOConfig.unsloth_num_chunks`) to a divisor of the batch size, but the `UnslothEfficientGRPO.apply` call site always passes the literal `1`, so the snapped value was never used. Worse, `UnslothEfficientGRPO.forward`'s uniform `div_(n_chunks)` is only correct at `n_chunks == 1`:

- Globally sum-normalized types (`dapo`/`cispo`/`vespo`) already divide by a global normalizer per chunk, so summing chunks is exact and a further `/n_chunks` makes the loss `n_chunks` times too small.
- Token-count-normalized `bnpo` needs the global `mask.sum()` as its normalizer.
- Any non-divisor chunk count mis-weights every loss type.

**Fix: option (a) chosen.** Removed the dead snapping code and documented that `n_chunks` is fixed at `1`, rather than shipping re-enabled chunking. During review, option (b) (per-loss-type global normalizers threaded through each chunk) was fully implemented and validated, but measurement showed chunking buys nothing today: since the logps-refactor the function receives MB-scale per-token log probabilities (the GB-scale logit memory is handled upstream by hidden-state return plus chunked and offloaded log-softmax), so peak memory is identical at any chunk count while per-row chunking is about 70 percent slower for this op. The caller-side chunk selection was therefore removed again in favor of the simpler single-chunk path. To be precise about scope: `UnslothEfficientGRPO` still contains its chunk loop and `div_(n_chunks)`, it is simply always called with `1`, and no in-tree caller can reach any other value. The `n_chunks` parameter is kept in the signature because the caller passes it as a keyword (`GRPOConfig.unsloth_num_chunks`), and a new warn-once helper `_warn_deprecated_n_chunks` tells users who set a non-default value that the option has no effect and points them at `unsloth_grpo_mini_batch` and `unsloth_logit_chunk_multiplier`, which control the memory that actually matters. This matches the theme of Issue 1: no silently ignored options. The `n_chunks=1` production path is numerically unchanged. The unused `numpy` import (only used by the removed snapping) was dropped.

## Verification

Added tests in `tests/test_rl_replacements_cpu.py`:

- The warning fires exactly once for each non-default option and lists both when both are set; it stays silent on the TRL defaults and when the attributes are missing (older TRL); and it fires at most once per trainer.
- `UnslothEfficientGRPO` at `n_chunks=1` is numerically identical (loss and gradient) to a direct autograd pass through `grpo_compute_loss` for all eight loss types: `grpo`, `bnpo`, `dr_grpo`, `dapo`, `cispo`, `sapo`, `vespo`, and `luspo`.
- The deprecation warning fires exactly once per process for non-default `unsloth_num_chunks` values, stays silent on the defaults (`-1`, `1`, `None`), and `grpo_accumulated_loss` provably never forwards `n_chunks` to `UnslothEfficientGRPO.apply`.

All 34 tests in `tests/test_rl_replacements_cpu.py` pass, and the two `rl_replacements` source-pattern regression tests in `test_zoo_history_regressions*.py` still pass.

Additionally confirmed on GPU (NVIDIA B200) with `torch.compile` enabled (the real production kernel path) that `UnslothEfficientGRPO` at `n_chunks=1` matches the naive `grpo_compute_loss` (loss and gradient) across all loss types in fp32.

## Issue 3: use_bias_correction_kl implemented

Rather than only warning, the optimized path now implements the option. Verified against TRL's `grpo_trainer.py` in both 0.27.0 and current main: when `beta != 0`, the per-token KL is multiplied by the pre-clamp, non-detached `coef_1` before the `loss_type` dispatch, at both token and sequence importance-sampling levels, and the corrected KL feeds both the `beta` term and the `kl` metric. `grpo_accumulated_loss` reads the flag with `getattr(trainer.args, "use_bias_correction_kl", False)`, so Unsloth follows whatever TRL itself would do. The field was added in TRL **0.26.0** and TRL flipped its own default from `False` to `True` in **1.10.0**. So on TRL up to 1.9.x, and on older TRL with no such field, existing runs are bit-identical, and on TRL 1.10.0 and newer the loss and gradients change by default, which is the correct TRL-parity behavior. Setting `use_bias_correction_kl=False` explicitly reproduces the old numbers exactly on every version. Covered by a TRL-formula oracle test (loss, gradient through the non-detached ratio, and `mean_kl`), flag-off bit-identity regressions, and an end-to-end `UnslothEfficientGRPO` pass across all eight loss types; 50 tests total pass.

---
## Verification of this PR against current main

Rebased onto `main` (the merge resolves the import block in favor of main's absolute
imports, and keeps both sides' tests). 78 tests pass in `tests/test_rl_replacements_cpu.py`.

**Numerical regression, 256 cases.** 8 loss types x {token, sequence} importance sampling
x beta in {0.04, 0} x 4 shapes (including `1x1`, `4x129`, ragged masks) x {float64, float32},
dumped from `origin/main` and from this branch and compared byte for byte:

| Comparison | Result |
|---|---|
| main vs this PR, `use_bias_correction_kl` off | 256/256 bit-identical, 0 errors, 0 non-finite |
| this PR, flag off vs on | 128 differ, 128 identical, and the 128 identical are exactly the `beta == 0` cases |

**TRL version matrix**, five isolated `uv venv`s, each importing a real `GRPOConfig`:

| TRL | field present | its default | what Unsloth forwards | result |
|---|---|---|---|---|
| 0.25.1 | no | n/a | `False` | bit-identical to pre-PR main |
| 0.27.0 | yes | `False` | `False` | bit-identical to pre-PR main |
| 1.9.0 | yes | `False` | `False` | bit-identical to pre-PR main |
| 1.10.0 | yes | `True` | `True` | corrected KL, matches the TRL formula |
| 1.12.0 | yes | `True` | `True` | corrected KL, matches the TRL formula |

An explicit `use_bias_correction_kl=False` reproduces the pre-PR numbers on all five, and
`top_entropy_quantile` defaults to `1.0` on all five, so nothing warns on defaults.

**Version skew**, since unsloth text-copies these bodies into
`unsloth_compiled_cache/UnslothGRPOTrainer.py`: a new cache body run against an
`unsloth_zoo` with neither helper, against an `unsloth_zoo` that cannot be imported at all,
an old pre-PR cache body (which carries its own `numpy` import) run against this branch,
and an old unsloth passing `unsloth_num_chunks` as a keyword. All are no-ops or work
unchanged; `grpo_accumulated_loss`'s signature is byte-identical to main's.

**GPU, NVIDIA B200, compiled production path**, four short GRPO runs on Qwen3-0.6B:

| Run | Result |
|---|---|
| default | finite losses; generated cache contains both helper imports; `apply` passes the literal `1` |
| `unsloth_num_chunks=4` | exactly one deprecation warning; losses, `kl` and grad-norms bit-identical to the default run |
| `top_entropy_quantile=0.2` | exactly one unsupported-option warning; training unchanged |
| `use_bias_correction_kl=True` | runs clean; grad-norm moves while the loss does not, which is the differentiable-ratio effect the option exists for |

The changed code contains no OS, path, or device specific constructs, and Unsloth Studio
does not reference the GRPO path at all.

---
## Dependency compatibility

**TRL, 16 versions in isolated `uv venv`s**, each importing a real `GRPOConfig` and running
`grpo_compute_loss` against the pre-PR dumps. Every row passes:

| TRL | field | its default | Unsloth forwards | numerics |
|---|---|---|---|---|
| 0.22.2, 0.23.1, 0.24.0, 0.25.1 | absent | n/a | `False` | bit-identical to pre-PR |
| 0.26.2, 0.27.0, 0.28.0, 0.29.1, 1.0.0, 1.3.0, 1.6.0, 1.7.1, 1.9.0 | yes | `False` | `False` | bit-identical to pre-PR |
| 1.10.0, 1.11.0, 1.12.0 | yes | `True` | `True` | corrected, matches the TRL formula |

An explicit `use_bias_correction_kl=False` reproduces the pre-PR numbers on all 16, and
`top_entropy_quantile` defaults to `1.0` on all 16, so nothing warns on defaults.

**transformers:** 4.57.6 and 5.0.0 / 5.8.0 / 5.16.1, each with TRL 1.12.0, plus 4.57.6 with
TRL 0.22.2. `GRPOConfig` constructs and the plumbing forwards TRL's own value in every combination.

**vLLM / `fast_inference = True`:** the vLLM importance sampling ratio scales the policy term
only. TRL applies it at `grpo_trainer.py:3236` and adds `beta * per_token_kl` at `:3239`;
`grpo_compute_loss` has the same ordering, so the corrected KL is never scaled by the sampling
ratio. Pinned by a new test over both `sequence_mask` and `token_truncate` modes, flag on and
off, checking loss, `mean_kl` and gradient against a TRL-formula oracle.

**torch:** the diff adds no `torch.*` call at all. The only tensor operation is
`kl_i = kl_i * coef_1`, a broadcasting multiply; everything else is `getattr`, a dict lookup,
f-strings and `logging`. `torch.compile(fullgraph=True)` over `grpo_compute_loss` with the flag
on and off matches eager in loss and gradient and raises no graph break, so the new branch stays
static.

**PEFT:** no PEFT symbol appears in the diff. The loss operates on per-token log probabilities
produced after the model forward, so the adapter implementation sits upstream of it. Exercised
end to end through `get_peft_model` on the GPU runs.

**Platform:** no `os`, `sys.platform`, `platform`, path or device construct in the diff, and no
MLX surface, so Linux, macOS, Windows and WSL run identical code and the NVIDIA / AMD / Intel /
MLX paths are untouched. Cross-platform CI (ubuntu-latest, macos-15, windows-latest) is running
on a staging repo to confirm. No browser surface: this PR does not touch Unsloth Studio, which
does not reference the GRPO path at all.

### CODE REVIEW

File: tests/test_rl_replacements_cpu.py [hunk 1]
```diff
@@ -210,3 +216,262 @@ def test_RL_REPLACEMENTS_contains_public_api_keys():
...
+
+    assert torch.allclose(
+        out[0].detach().double(), loss_ref.detach().double(), atol=1e-8, rtol=1e-6
+    ), f"{loss_type}: loss mismatch"
+    assert torch.allclose(
+        new_eff.grad, new_ref.grad, atol=1e-8, rtol=1e-6
+    ), f"{loss_type}: gradient mismatch"
+
+
+_ALL_LOSS_TYPES = ["grpo", "bnpo", "dr_grpo", "dapo", "cispo", "sapo", "luspo"]
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply tests/test_rl_replacements_cpu.py]: vespo was not actually in the parametrization (the loss_type list was grpo/bnpo/dr_grpo/dapo/cispo/sapo/luspo), so no cases were failing and the suite passed. But the gap it points at is real, so I added it in 7c6d1c58: the fixture now supplies a faithful per-sequence get_gamma_weights (mirroring TRL GRPOTrainer.get_gamma_weights) for vespo, and vespo is in the invariance and ragged-chunk parametrizations. All eight loss types are now proven chunk-invariant (gradient bit-identical, loss/metrics at fp32 resolution). vespo shares the global cispo/dapo normalizer and its gamma weights are per-sequence, so it chunks safely.

**Comment by danielhanchen** [comment]
Update: commit c23d3225 changes direction on the chunking half of this PR. The global-normalizer chunking implementation was correct, but benchmarking showed it saves no memory on the current code path (the function receives per-token log probabilities, which are MB scale; the GB scale logit memory is already handled upstream by the hidden state return plus chunked and offloaded log softmax) while per row chunking is about 70 percent slower for the op. So the chunking machinery is removed and the call stays whole batch, matching the original intent of this PR. unsloth_num_chunks is still accepted for compatibility and now emits a one time deprecation warning pointing at unsloth_grpo_mini_batch and unsloth_logit_chunk_multiplier. The unsupported options guard is unchanged. 34 tests pass.

**Comment by danielhanchen** [comment]
Commit d4175e53 implements use_bias_correction_kl instead of warning that it is unsupported. The correction matches TRL exactly (verified against grpo_trainer.py in both TRL 0.27.0 and current main): inside the beta != 0 branch, per token KL is multiplied by the pre clamp, non detached coef_1 before the loss_type dispatch, at both token and sequence importance sampling levels, and the corrected KL feeds both the beta term and the kl metric. The flag is read with getattr and defaults to False, so older TRL without the field and all existing runs are bit identical. _warn_unsupported_grpo_options now only covers top_entropy_quantile. 16 new tests including a TRL formula oracle for loss, gradient, and mean_kl; 50 total pass.

**Comment by danielhanchen** [comment]
Commit 2e3cb951 is a comment-only cleanup (verified AST-identical code, tests unchanged at 50 passing).

**Comment by danielhanchen** [comment]
Merged current main into this branch and corrected two claims that no longer held.

The branch had drifted behind main and conflicted in both files (main's absolute imports and `torch_compile_with_fallback` wiring landed in the same regions). The merge keeps main's import block, spells the new `logger` import absolutely to match, and keeps both sides' tests. 78 tests pass.

Two corrections in `e4c8758`:

- TRL flipped `use_bias_correction_kl`'s own default from `False` to `True` in **1.10.0** (`False` in 0.27.0, 1.7.1, 1.8.0, 1.9.0; `True` in 1.10.0, 1.11.0, 1.12.0). The code was already correct, since `grpo_accumulated_loss` reads the trainer's actual value, but "every existing run stays bit-identical" was only true up to TRL 1.9.x. On 1.10.0 and newer the loss and gradients change by default, which is exactly TRL parity. The docstrings and the PR description now say so, and an explicit `use_bias_correction_kl=False` still reproduces the old numbers on every version.
- The deprecation message said loss chunking was removed. What was removed is the caller-side chunk selection. `UnslothEfficientGRPO` still has its chunk loop and is simply always called with `1`. The message says that now.

Also, `_warn_unsupported_grpo_options` now latches its per-trainer flag only after it has actually warned, so a config changed mid-run still gets its one warning, and the latch write is guarded for trainers that do not accept new attributes.

Verification added to the description: 256 numerical cases bit-identical to main with the flag off, a five-version TRL matrix in isolated venvs, version-skew simulations for the generated trainer cache, and four GPU runs on the compiled path.

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex security review

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex review
